[Bug]: 500 error when trying to log in to dashboard

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Deploy type

Downstream version (eg. `RHODS 1.29`)

### Version

2.4.0

### Current Behavior

On a cluster that has 

There is this error in the oauth-proxy container of the rhods dashboard pods:

```
2023/11/07 14:19:35 provider.go:631: Performing OAuth discovery against https://172.30.0.1/.well-known/oauth-authorization-server
2023/11/07 14:19:35 provider.go:671: 200 GET https://172.30.0.1/.well-known/oauth-authorization-server {
"issuer": "https://oauth-openshift.apps.gcp-4-13-2-4.mzx2.s2.devshift.org/",
"authorization_endpoint": "https://oauth-openshift.apps.gcp-4-13-2-4.mzx2.s2.devshift.org/oauth/authorize",
"token_endpoint": "https://oauth-openshift.apps.gcp-4-13-2-4.mzx2.s2.devshift.org/oauth/token",
"scopes_supported": [
"user:check-access",
"user:full",
"user:info",
"user:list-projects",
"user:list-scoped-projects"
],
"response_types_supported": [
"code",
"token"
],
"grant_types_supported": [
"authorization_code",
"implicit"
],
"code_challenge_methods_supported": [
"plain",
"S256"
]
}
2023/11/07 14:19:35 oauthproxy.go:654: error redeeming code (client:10.129.2.8:52702): got 400 from "https://oauth-openshift.apps.gcp-4-13-2-4.mzx2.s2.devshift.org/oauth/token" {"error":"unauthorized_client","error_description":"The client is not authorized to request a token using this method."}
2023/11/07 14:19:35 oauthproxy.go:445: ErrorPage 500 Internal Error Internal Error
```

### Expected Behavior

It should be possible to log into the dashboard seamlessly when upgrading or installing RHODS 2.4 without any disabling/enabling process

### Steps To Reproduce

1) clean the cluster with: https://gitlab.cee.redhat.com/data-hub/olminstall/-/blob/main/cleanup.sh?ref_type=heads
2) manually remove:
2.1) [https://console-openshift-console.apps.ods-qe-03.rhods.ccitredhat.com/k8s/cluster/customresourcedefinitions[…]tions.dscinitialization.opendatahub.io](https://console-openshift-console.apps.ods-qe-03.rhods.ccitredhat.com/k8s/cluster/customresourcedefinitions/dscinitializations.dscinitialization.opendatahub.io)
2.2) [https://console-openshift-console.apps.ods-qe-03.rhods.ccitredhat.com/k8s/cluster/customresourcedefinitions[…]ters.datasciencecluster.opendatahub.io](https://console-openshift-console.apps.ods-qe-03.rhods.ccitredhat.com/k8s/cluster/customresourcedefinitions/datascienceclusters.datasciencecluster.opendatahub.io)
3) install RHODS 2.4
4) you will see that DSC Initialization is present
5) create a Data Science Cluster
you will have the error above



### Workaround (if any)

Disabling the dashboard component from the DSC and enabling it again fixed the problem

--- a/components/codeflare/codeflare.go
+++ b/components/codeflare/codeflare.go
@@ -51,7 +51,7 @@ func (c *CodeFlare) GetComponentName() string {
 // Verifies that CodeFlare implements ComponentInterface.
 var _ components.ComponentInterface = (*CodeFlare)(nil)
 
-func (c *CodeFlare) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec) error {
+func (c *CodeFlare) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec, _ bool) error {
 	var imageParamMap = map[string]string{
 		"odh-codeflare-operator-controller-image": "RELATED_IMAGE_ODH_CODEFLARE_OPERATOR_IMAGE", // no need mcad, embedded in cfo
 		"namespace": dscispec.ApplicationsNamespace,

--- a/components/component.go
+++ b/components/component.go
@@ -75,7 +75,7 @@ type ManifestsConfig struct {
 }
 
 type ComponentInterface interface {
-	ReconcileComponent(cli client.Client, owner metav1.Object, DSCISpec *dsciv1.DSCInitializationSpec) error
+	ReconcileComponent(cli client.Client, owner metav1.Object, DSCISpec *dsciv1.DSCInitializationSpec, currentComponentStatus bool) error
 	Cleanup(cli client.Client, DSCISpec *dsciv1.DSCInitializationSpec) error
 	GetComponentName() string
 	GetManagementState() operatorv1.ManagementState

--- a/components/dashboard/dashboard.go
+++ b/components/dashboard/dashboard.go
@@ -5,13 +5,13 @@ package dashboard
 import (
 	"context"
 	"fmt"
-	v1 "k8s.io/api/core/v1"
-	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	"path/filepath"
 	"strings"
 
 	operatorv1 "github.com/openshift/api/operator/v1"
 	routev1 "github.com/openshift/api/route/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
@@ -75,6 +75,7 @@ func (d *Dashboard) GetComponentName() string {
 // Verifies that Dashboard implements ComponentInterface.
 var _ components.ComponentInterface = (*Dashboard)(nil)
 
+//nolint:gocyclo
 func (d *Dashboard) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec, currentComponentStatus bool) error {
 	var imageParamMap = map[string]string{
 		"odh-dashboard-image": "RELATED_IMAGE_ODH_DASHBOARD_IMAGE",
@@ -89,7 +90,6 @@ func (d *Dashboard) ReconcileComponent(cli client.Client, owner metav1.Object, d
 
 	// Update Default rolebinding
 	if enabled {
-
 		if err := d.cleanOauthClientSecrets(cli, dscispec, currentComponentStatus); err != nil {
 			return err
 		}

--- a/components/dashboard/dashboard.go
+++ b/components/dashboard/dashboard.go
@@ -5,6 +5,8 @@ package dashboard
 import (
 	"context"
 	"fmt"
+	v1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	"path/filepath"
 	"strings"
 
@@ -73,7 +75,7 @@ func (d *Dashboard) GetComponentName() string {
 // Verifies that Dashboard implements ComponentInterface.
 var _ components.ComponentInterface = (*Dashboard)(nil)
 
-func (d *Dashboard) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec) error {
+func (d *Dashboard) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec, currentComponentStatus bool) error {
 	var imageParamMap = map[string]string{
 		"odh-dashboard-image": "RELATED_IMAGE_ODH_DASHBOARD_IMAGE",
 	}
@@ -87,6 +89,10 @@ func (d *Dashboard) ReconcileComponent(cli client.Client, owner metav1.Object, d
 
 	// Update Default rolebinding
 	if enabled {
+
+		if err := d.cleanOauthClientSecrets(cli, dscispec, currentComponentStatus); err != nil {
+			return err
+		}
 		// Download manifests and update paths
 		if err := d.OverrideManifests(string(platform)); err != nil {
 			return err
@@ -257,5 +263,30 @@ func (d *Dashboard) deployConsoleLink(cli client.Client, owner metav1.Object, na
 		return fmt.Errorf("failed to set dashboard consolelink from %s: %w", PathConsoleLink, err)
 	}
 
+	return nil
+}
+
+func (d *Dashboard) cleanOauthClientSecrets(cli client.Client, dscispec *dsciv1.DSCInitializationSpec, currentComponentStatus bool) error {
+	// Remove previous oauth-client secrets
+	// Check if component is going from state of `Not Installed --> Installed`
+	// Assumption: Component is currently set to enabled
+	if !currentComponentStatus {
+		// Delete client secrets from previous installation
+		oauthClientSecret := &v1.Secret{}
+		err := cli.Get(context.TODO(), client.ObjectKey{
+			Namespace: dscispec.ApplicationsNamespace,
+			Name:      "dashboard-oauth-client",
+		}, oauthClientSecret)
+		if err != nil {
+			if !apierrs.IsNotFound(err) {
+				return err
+			}
+		} else {
+			err := cli.Delete(context.TODO(), oauthClientSecret)
+			if err != nil {
+				return fmt.Errorf("error deleting oauth client secret: %v", err)
+			}
+		}
+	}
 	return nil
 }

--- a/components/datasciencepipelines/datasciencepipelines.go
+++ b/components/datasciencepipelines/datasciencepipelines.go
@@ -48,7 +48,7 @@ func (d *DataSciencePipelines) GetComponentName() string {
 // Verifies that Dashboard implements ComponentInterface.
 var _ components.ComponentInterface = (*DataSciencePipelines)(nil)
 
-func (d *DataSciencePipelines) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec) error {
+func (d *DataSciencePipelines) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec, _ bool) error {
 	var imageParamMap = map[string]string{
 		"IMAGES_APISERVER":         "RELATED_IMAGE_ODH_ML_PIPELINES_API_SERVER_IMAGE",
 		"IMAGES_ARTIFACT":          "RELATED_IMAGE_ODH_ML_PIPELINES_ARTIFACT_MANAGER_IMAGE",

--- a/components/kserve/kserve.go
+++ b/components/kserve/kserve.go
@@ -72,7 +72,7 @@ func (k *Kserve) GetComponentName() string {
 // Verifies that Kserve implements ComponentInterface.
 var _ components.ComponentInterface = (*Kserve)(nil)
 
-func (k *Kserve) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec) error {
+func (k *Kserve) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec, _ bool) error {
 	// paramMap for Kserve to use.
 	var imageParamMap = map[string]string{}
 

--- a/components/modelmeshserving/modelmeshserving.go
+++ b/components/modelmeshserving/modelmeshserving.go
@@ -2,7 +2,6 @@
 package modelmeshserving
 
 import (
-	"context"
 	"path/filepath"
 	"strings"
 
@@ -134,15 +133,6 @@ func (m *ModelMeshServing) ReconcileComponent(cli client.Client, owner metav1.Ob
 		} else {
 			return err
 		}
-	}
-
-	// Get monitoring namespace
-	dscInit := &dsciv1.DSCInitialization{}
-	err = cli.Get(context.TODO(), client.ObjectKey{
-		Name: "default",
-	}, dscInit)
-	if err != nil {
-		return err
 	}
 
 	// CloudService Monitoring handling

--- a/components/modelmeshserving/modelmeshserving.go
+++ b/components/modelmeshserving/modelmeshserving.go
@@ -67,7 +67,7 @@ func (m *ModelMeshServing) GetComponentName() string {
 // Verifies that Dashboard implements ComponentInterface.
 var _ components.ComponentInterface = (*ModelMeshServing)(nil)
 
-func (m *ModelMeshServing) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec) error {
+func (m *ModelMeshServing) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec, _ bool) error {
 	var imageParamMap = map[string]string{
 		"odh-mm-rest-proxy":             "RELATED_IMAGE_ODH_MM_REST_PROXY_IMAGE",
 		"odh-modelmesh-runtime-adapter": "RELATED_IMAGE_ODH_MODELMESH_RUNTIME_ADAPTER_IMAGE",

--- a/components/ray/ray.go
+++ b/components/ray/ray.go
@@ -48,7 +48,7 @@ func (r *Ray) GetComponentName() string {
 // Verifies that Ray implements ComponentInterface.
 var _ components.ComponentInterface = (*Ray)(nil)
 
-func (r *Ray) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec) error {
+func (r *Ray) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec, _ bool) error {
 	var imageParamMap = map[string]string{
 		"odh-kuberay-operator-controller-image": "RELATED_IMAGE_ODH_KUBERAY_OPERATOR_CONTROLLER_IMAGE",
 		"namespace":                             dscispec.ApplicationsNamespace,

--- a/components/trustyai/trustyai.go
+++ b/components/trustyai/trustyai.go
@@ -47,7 +47,7 @@ func (t *TrustyAI) GetComponentName() string {
 // Verifies that TrustyAI implements ComponentInterface.
 var _ components.ComponentInterface = (*TrustyAI)(nil)
 
-func (t *TrustyAI) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec) error {
+func (t *TrustyAI) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsciv1.DSCInitializationSpec, _ bool) error {
 	var imageParamMap = map[string]string{
 		"trustyaiServiceImage":  "RELATED_IMAGE_ODH_TRUSTYAI_SERVICE_IMAGE",
 		"trustyaiOperatorImage": "RELATED_IMAGE_ODH_TRUSTYAI_SERVICE_OPERATOR_IMAGE",

--- a/components/workbenches/workbenches.go
+++ b/components/workbenches/workbenches.go
@@ -92,7 +92,7 @@ func (w *Workbenches) GetComponentName() string {
 // Verifies that Workbench implements ComponentInterface.
 var _ components.ComponentInterface = (*Workbenches)(nil)
 
-func (w *Workbenches) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsci.DSCInitializationSpec) error {
+func (w *Workbenches) ReconcileComponent(cli client.Client, owner metav1.Object, dscispec *dsci.DSCInitializationSpec, _ bool) error {
 	var imageParamMap = map[string]string{
 		"odh-notebook-controller-image":    "RELATED_IMAGE_ODH_NOTEBOOK_CONTROLLER_IMAGE",
 		"odh-kf-notebook-controller-image": "RELATED_IMAGE_ODH_KF_NOTEBOOK_CONTROLLER_IMAGE",

--- a/controllers/datasciencecluster/datasciencecluster_controller.go
+++ b/controllers/datasciencecluster/datasciencecluster_controller.go
@@ -271,7 +271,7 @@ func (r *DataScienceClusterReconciler) reconcileSubComponent(ctx context.Context
 	}
 
 	// Reconcile component
-	err = component.ReconcileComponent(r.Client, instance, r.DataScienceCluster.DSCISpec)
+	err = component.ReconcileComponent(r.Client, instance, r.DataScienceCluster.DSCISpec, instance.Status.InstalledComponents[componentName])
 
 	if err != nil {
 		// reconciliation failed: log errors, raise event and update status accordingly

--- a/controllers/secretgenerator/secretgenerator_controller.go
+++ b/controllers/secretgenerator/secretgenerator_controller.go
@@ -19,6 +19,7 @@ package secretgenerator
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 	"time"
 
@@ -227,8 +228,16 @@ func (r *SecretGeneratorReconciler) createOAuthClient(ctx context.Context, name 
 	err := r.Client.Create(ctx, oauthClient)
 	if err != nil {
 		if apierrs.IsAlreadyExists(err) {
-			secGenLog.Info("OAuth client resource already exists", "name", oauthClient.Name)
-
+			secGenLog.Info("OAuth client resource already exists, patch it", "name", oauthClient.Name)
+			data, err := json.Marshal(oauthClient)
+			if err != nil {
+				return fmt.Errorf("failed to get DataScienceCluster custom resource data: %v", err)
+			}
+			err = r.Client.Patch(context.TODO(), oauthClient, client.RawPatch(types.ApplyPatchType, data),
+				client.ForceOwnership, client.FieldOwner("opendatahub-operator"))
+			if err != nil {
+				return err
+			}
 			return nil
 		}
 	}

--- a/main.go
+++ b/main.go
@@ -180,14 +180,12 @@ func main() { //nolint:funlen
 	if !disableDSCConfig {
 		if err = upgrade.CreateDefaultDSCI(setupClient, platform, dscApplicationsNamespace, dscMonitoringNamespace); err != nil {
 			setupLog.Error(err, "unable to create initial setup for the operator")
-			os.Exit(1)
 		}
 	}
 
 	// Apply update from legacy operator
 	if err = upgrade.UpdateFromLegacyVersion(setupClient, platform); err != nil {
 		setupLog.Error(err, "unable to update from legacy operator version")
-		os.Exit(1)
 	}
 
 	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {

--- a/pkg/upgrade/upgrade.go
+++ b/pkg/upgrade/upgrade.go
@@ -220,6 +220,14 @@ func CreateDefaultDSCI(cli client.Client, platform deploy.Platform, appNamespace
 		Spec: *defaultDsciSpec,
 	}
 
+	patchedDSCI := &dsci.DSCInitialization{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "DSCInitialization",
+			APIVersion: "dscinitialization.opendatahub.io/v1",
+		},
+		Spec: *defaultDsciSpec,
+	}
+
 	instances := &dsci.DSCInitializationList{}
 	if err := cli.List(context.TODO(), instances); err != nil {
 		return err
@@ -227,14 +235,16 @@ func CreateDefaultDSCI(cli client.Client, platform deploy.Platform, appNamespace
 
 	switch {
 	case len(instances.Items) > 1:
-		return fmt.Errorf("only one instance of DSCInitialization object is allowed. Please delete other instances ")
+		fmt.Printf("only one instance of DSCInitialization object is allowed. Please delete other instances ")
+		return nil
 	case len(instances.Items) == 1:
 		if platform == deploy.ManagedRhods || platform == deploy.SelfManagedRhods {
-			data, err := json.Marshal(defaultDsciSpec)
+			data, err := json.Marshal(patchedDSCI)
 			if err != nil {
 				return err
 			}
-			err = cli.Patch(context.TODO(), defaultDsci, client.RawPatch(types.ApplyPatchType, data),
+			existingDSCI := &instances.Items[0]
+			err = cli.Patch(context.TODO(), existingDSCI, client.RawPatch(types.ApplyPatchType, data),
 				client.ForceOwnership, client.FieldOwner("opendatahub-operator"))
 			if err != nil {
 				return err

